src/parsing/prscParser.ts

```diff
@@ -0,0 +1,2795 @@
+	);
+
+	const contextItemDecl: Parser<IAST> = then(
+		precededMultiple(
+			[token('declare'), whitespacePlus, token('context'), whitespacePlus, token('item')],
+			optional(precededMultiple([whitespacePlus, token('as')], itemType))
+		),
+		or([
+			map(preceded(surrounded(token(':='), whitespace), varValue), (x) => ['varValue', x]),
+			map(
+				precededMultiple(
+					[whitespacePlus, token('external')],
+					optional(preceded(surrounded(token(':='), whitespace), varDefaultValue))
+				),
+				(_) => ['external']
+			),
+		]),
+		(itemTypePart, val) =>
+			[
+				'contextItemDecl',
+				...(itemTypePart ? [['contextItemType', itemTypePart]] : []),
+				val,
+			] as IAST
+	);
+
+	const prolog: Parser<IAST> = then(
+		star(
+			followed(
+				or([defaultNamespaceDecl, setter, namespaceDecl, importExpr]),
+				surrounded(separator, whitespace)
+			)
+		),
+		star(
+			followed(
+				or([contextItemDecl, annotatedDecl, optionDecl]),
+				surrounded(separator, whitespace)
+			)
+		),
+		(moduleSettings, declarations) =>
+			moduleSettings.length === 0 && declarations.length === 0
+				? null
+				: ['prolog', ...moduleSettings, ...declarations]
+	);
+
+	const moduleDecl: Parser<IAST> = precededMultiple(
+		[token('module'), whitespacePlus, token('namespace'), whitespacePlus],
+		then(
+			followed(ncName, surrounded(token('='), whitespace)),
+			followed(uriLiteral, surrounded(separator, whitespace)),
+			(prefixPart, uri) => ['moduleDecl', ['prefix', prefixPart], ['uri', uri]]
+		)
+	);
+
+	const libraryModule: Parser<IAST> = then(
+		moduleDecl,
+		preceded(whitespace, prolog),
+		(moduleDeclPart, prologPart) =>
+			['libraryModule', moduleDeclPart, ...(prologPart ? [prologPart] : [])] as IAST
+	);
+
+	const mainModule: Parser<IAST> = then(
+		prolog,
+		preceded(whitespace, queryBody),
+		(prologPart, body) => ['mainModule', ...(prologPart ? [prologPart] : []), body]
+	);
+
+	const versionDeclaration: Parser<IAST> = map(
+		precededMultiple(
+			[token('xquery'), whitespace],
+			followed(
+				or([
+					then(
+						preceded(token('encoding'), whitespacePlus),
+						stringLiteral,
+						(encoding) => ['encoding', encoding] as IAST | IAST[]
+					),
+					then(
+						precededMultiple([token('version'), whitespacePlus], stringLiteral),
+						optional(
+							precededMultiple(
+								[whitespacePlus, token('encoding'), whitespacePlus],
+								stringLiteral
+							)
+						),
+						(version, encoding) =>
+							[
+								['version', version],
+								...(encoding ? [['encoding', encoding]] : []),
+							] as IAST[]
+					),
+				]),
+				preceded(whitespace, separator)
+			)
+		),
+		(x) => ['versionDecl', ...x]
+	);
+
+	const module: Parser<IAST> = then(
+		optional(surrounded(versionDeclaration, whitespace)),
+		or([libraryModule, mainModule]),
+		(versionDecl, modulePart) =>
+			['module', ...(versionDecl ? [versionDecl] : []), ...[modulePart]] as IAST
```

Comment:
Cool to see all these tokens in TypeScript now. Maybe in the future we can something akin to what y'all did earlier: replace the strings with number constants / enums. While parsing is not a significant part of the performance of the engine, it might remove the risk of some silly typos in silly places.

Not for now though!